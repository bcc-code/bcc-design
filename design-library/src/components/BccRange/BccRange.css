@layer components {
    .bcc-range {
        @apply flex flex-col gap-2 p-8 w-full;
    }

    .bcc-range__container {
        @apply flex flex-col items-center gap-2 w-full relative;
    }

    .bcc-range__text {
        @apply text-sm text-gray-700 whitespace-nowrap;
    }

    .bcc-range__value {
        @apply flex absolute bottom-4 -translate-x-1/2 bg-danger mb-2 text-sm rounded-lg w-12 h-8 text-center items-center justify-center shadow-md;
    }

    .bcc-range__input {
        @apply w-full h-1 bg-gray-200 rounded appearance-none cursor-pointer;
        @apply disabled:opacity-50 disabled:cursor-not-allowed;
        -webkit-appearance: none;
        appearance: none;
    }

    .bcc-range__input:disabled {
        @apply bg-gray-400 cursor-not-allowed;
    }

    .bcc-range__input::-webkit-slider-thumb {
        -webkit-appearance: none;
        @apply w-5 h-5 mt-[-8px] rounded-full bg-white cursor-pointer;
        box-shadow: 0 1px 2px 0px rgba(113, 113, 113, 0.5),
            0 0 0 2px rgba(226, 232, 240, 1),
            0 0 0 4px rgb(255, 255, 255),
            0 0 0 5px rgba(62, 142, 117, 1);
    }

    .bcc-range__input::-webkit-slider-runnable-track {
        @apply h-1 rounded bg-gray-200;
    }

    .bcc-range__input::-moz-range-thumb {
        @apply w-4 h-4 rounded-full bg-white cursor-pointer border-0;
        box-shadow: 0 1px 2px 0px rgba(113, 113, 113, 0.5),
            0 0 0 2px rgba(226, 232, 240, 1),
            0 0 0 4px rgb(255, 255, 255),
            0 0 0 5px rgba(62, 142, 117, 1);
    }

    .bcc-range__input::-moz-range-track {
        @apply h-1 rounded bg-gray-200;
    }

}