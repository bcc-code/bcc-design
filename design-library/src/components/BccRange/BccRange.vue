<script setup lang="ts">
import { onMounted, computed } from "vue";

const props = defineProps({
  min: { type: Number, default: 0 },
  max: { type: Number, default: 10 },
  step: { type: Number, default: 1 },
  disabled: { type: Boolean, default: false },
  fromLeft: { type: Boolean, default: true },
});

const value = defineModel<number>({ required: true });

const handleInput = (event: Event) => {
  const target = event.target as HTMLInputElement;
  value.value = Number(target.value);
};

onMounted(() => {
  value.value = props.fromLeft ? props.min : Math.round((props.max - props.min) / 2);
});

const valuePosition = computed(() => {
  const percentage = ((value.value - props.min) / (props.max - props.min)) * 100;
  const thumbHalfWidth = 8;

  const leftEdgeProximity = Math.max(0, 1 - percentage / 70);
  const rightEdgeProximity = Math.max(0, 1 - (100 - percentage) / 70);

  const leftOffset = thumbHalfWidth * leftEdgeProximity;
  const rightOffset = -thumbHalfWidth * rightEdgeProximity;

  console.log(
    percentage,
    leftEdgeProximity,
    rightEdgeProximity,
    leftOffset,
    rightOffset,
    `calc(${percentage}% + ${leftOffset + rightOffset}px)`
  );
  return `calc(${percentage}% + ${leftOffset + rightOffset}px)`;
});

const valueColor = computed(() => {
  const thirds = (props.max - props.min) / 3;
  if (value.value < thirds) {
    return "bg-success text-success";
  } else if (value.value < thirds * 2) {
    return "bg-warning text-warning";
  } else {
    return "bg-danger text-danger";
  }
});
</script>

<template>
  <div class="bcc-range">
    <div class="bcc-range__container">
      <div class="bcc-range__value" :class="valueColor" :style="{ left: valuePosition }">
        {{ value }}
      </div>
      <input
        type="range"
        class="bcc-range__input"
        :min="min"
        :max="max"
        :step="step"
        :value="value"
        :disabled="disabled"
        @input="handleInput"
      />
    </div>
  </div>
</template>
