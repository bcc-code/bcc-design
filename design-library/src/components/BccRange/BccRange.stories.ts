import BccRange from "./BccRange.vue";

import type { Meta, StoryFn } from "@storybook/vue3";

export default {
  title: "Forms/BccRange",
  component: BccRange,
  argTypes: {
    modelValue: {
      description: "The current value of the range slider",
    },
    min: {
      description: "The minimum value of the range slider",
    },
    max: {
      description: "The maximum value of the range slider",
    },
    step: {
      description: "The step value of the range slider",
    },
    disabled: {
      description: "Whether the range slider is disabled",
    },
    fromLeft: {
      description: "Whether the range slider starts from the left or middle",
    },
  },
} as Meta<typeof BccRange>;

const Template: StoryFn<typeof BccRange> = (args) => ({
  components: { BccRange },
  setup() {
    return { args };
  },
  template: `
    <BccRange />
  `,
});

export const Example = Template.bind({});
Example.args = {
  modelValue: 0,
  min: -10,
  max: 10,
  step: 1,
  disabled: false,
};
Example.parameters = {
  docs: {
    source: {
      language: "html",
      code: `
    <BccRange />`,
    },
  },
};

export const Disabled: StoryFn<typeof BccRange> = () => ({
  components: { BccRange },
  template: `
    <BccRange disabled />
  `,
});

export const WithCustomRange: StoryFn<typeof BccRange> = () => ({
  components: { BccRange },
  template: `
    <BccRange min=0  max=100 step=2 />
  `,
});

export const StartFromMiddle: StoryFn<typeof BccRange> = () => ({
  components: { BccRange },
  template: `
    <BccRange :fromLeft=false min=0 max=10 />
  `,
});
