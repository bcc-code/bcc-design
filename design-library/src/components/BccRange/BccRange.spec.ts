import { describe, it, expect } from "vitest";
import { mount } from "@vue/test-utils";
import BccRange from "./BccRange.vue";

describe("BccRange", () => {
  it("renders a range input", () => {
    const wrapper = mount(BccRange, {
      props: { modelValue: 0, min: 0, max: 20, step: 1 },
    });
    expect(wrapper.html()).toMatchSnapshot();
  });
});
