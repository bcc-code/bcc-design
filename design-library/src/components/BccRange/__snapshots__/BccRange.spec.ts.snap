// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`BccRange > renders a range input 1`] = `
"<div class="bcc-range">
  <div class="bcc-range__container">
    <div class="bcc-range__value bg-success text-success" style="left: calc(0% + 8px);">0</div><input type="range" class="bcc-range__input" min="0" max="20" step="1">
  </div>
</div>"
`;
